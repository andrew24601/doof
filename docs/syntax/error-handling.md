# Error Handling with panic

Use `panic(message: string)` for unrecoverable errors. It prints the message to stderr and terminates the process (`std::exit(1)` in generated C++).

```doof
function divide(x: int, y: int): int {
    if (y == 0) { panic("Division by zero is not allowed"); }
    return x / y;
}

function validateInput(value: int): void {
    if (value < 0) { panic("Negative values are not supported"); }
}
```

Guidelines:
- `panic` never returns; use for unrecoverable errors
- Prefer optional types or error codes for recoverable cases
