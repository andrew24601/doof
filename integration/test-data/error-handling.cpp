#include "error-handling.h"
#include "doof_runtime.h"

int safeDivision(int x, int y) {
    if ((y == 0)) {
        (std::cerr << "panic: " << "Division by zero error" << std::endl, std::exit(1), 0);
    }
    return (x / y);
}

void validateInput(int value) {
    if ((value < 0)) {
        (std::cerr << "panic: " << "Negative values not allowed" << std::endl, std::exit(1), 0);
    }
}

int main() {
    int result1 = safeDivision(10, 2);
    std::cout << "Safe division result: 5" << std::endl;
    validateInput(5);
    std::cout << "Input validation passed" << std::endl;
    int result2 = safeDivision(10, 0);
    std::cout << "This should never be reached" << std::endl;
    return 0;
}

