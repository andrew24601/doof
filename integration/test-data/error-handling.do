// Test panic function for error handling

function safeDivision(x: int, y: int): int {
    if (y == 0) {
        panic("Division by zero error");
    }
    return x / y;
}

function validateInput(value: int): void {
    if (value < 0) {
        panic("Negative values not allowed");
    }
}

function main(): int {
    // This should work fine
    let result1 = safeDivision(10, 2);
    println("Safe division result: 5");
    
    // This should also work
    validateInput(5);
    println("Input validation passed");
    
    // This will panic and exit the program
    let result2 = safeDivision(10, 0);
    println("This should never be reached");
    
    return 0;
}