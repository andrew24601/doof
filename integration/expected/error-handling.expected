Safe division result: 5
Input validation passed
